### DISCUSSION TITLE
createJSONStorage() should allow to provide seralizer and deserializer

### DISCUSSION THREAD

**Discussion by macmillan78** [pmndrs/jotai#2119] [Ideas] [answered]
In our setup we use bigints for IDs, thus createJSONStorage() create a storage unable to deal with our IDs. Making it possible to provide a serializer and a deserializer would make it easy to create a storage capable to deal with bigints or any other unusual type of data.

Maybe it would be more consistent to just provide createSerializingStorage() with createJSONStorage() beeing a special kind of it.

**Comment by dai-shi** [comment, marked as answer]
Although it's not clear, our recommendation is not to use createJSONStorage. It's just an _easy solution_ with JSON.parse which can cause runtime error, and not type safe. We should probably improve the docs about the recommended usage.

All you need is to implement the SyncStorage interface (or the async one) and pass it to `atomWithStorage`.

https://github.com/pmndrs/jotai/blob/f7d7f74eb64660421dfea78d2be1e47816ffd6be/src/vanilla/utils/atomWithStorage.ts#L26-L35

**Reply by macmillan78** [reply]
Perfect, this is what we did. Thanks for the clarification!
